lib/Dialect/Comb/CombFolds.cpp

```diff
@@ -988,6 +988,188 @@ OpFoldResult ICmpOp::fold(ArrayRef<Attribute> constants) {
+  size_t totalWidth = 0;
+  for (auto operand : range) {
+    // getIntOrFloatBitWidth should never raise, since all arguments to ConcatOp
+    // are integers.
+    ssize_t width = operand.getType().getIntOrFloatBitWidth();
+    assert(width >= 0);
+    totalWidth += width;
+  }
+  return totalWidth;
+}
+
+static Optional<Value> findFirstNonEmptyValue(const OperandRange &range) {
+  for (auto op : range) {
+    if (op.getType().getIntOrFloatBitWidth() > 0) {
+      return mlir::Optional<Value>(op);
+    }
+  }
+
+  return None;
+}
+
+static bool isPredicateSigned(ICmpPredicate predicate) {
+  switch (predicate) {
+  case ICmpPredicate::ult:
+  case ICmpPredicate::ugt:
+  case ICmpPredicate::ule:
+  case ICmpPredicate::uge:
+  case ICmpPredicate::ne:
+  case ICmpPredicate::eq:
+    return false;
+  case ICmpPredicate::slt:
+  case ICmpPredicate::sgt:
+  case ICmpPredicate::sle:
+  case ICmpPredicate::sge:
+    return true;
+  }
+  llvm_unreachable("unknown comparison predicate");
+}
+
+// Reduce the strength icmp(concat(...), concat(...)) by doing a element-wise
+// comparison on common prefix and suffixes. Returns success() if a rewriting
+// happens.
+static LogicalResult matchAndRewriteCompareConcat(ICmpOp &op, ConcatOp &lhs,
+                                                  ConcatOp &rhs,
+                                                  PatternRewriter &rewriter) {
+  auto lhsOperands = lhs.getOperands();
+  auto rhsOperands = rhs.getOperands();
+  size_t numElements = std::min<size_t>(lhsOperands.size(), rhsOperands.size());
+
+  std::pair<size_t, size_t> commonPrefixSuffixLength =
+      computeCommonPrefixAndSuffix(lhsOperands, rhsOperands);
+  size_t commonPrefixLength = commonPrefixSuffixLength.first;
+  size_t commonSuffixLength = commonPrefixSuffixLength.second;
+
+  auto replaceWithConstantI1 = [&](bool constant) -> LogicalResult {
+    rewriter.replaceOpWithNewOp<hw::ConstantOp>(op, APInt(1, constant));
+    return success();
+  };
+
+  auto directOrCat = [&](const OperandRange &range) -> Value {
+    assert(range.size() > 0);
+    if (range.size() == 1) {
+      return range.front();
+    }
+
+    return rewriter.create<ConcatOp>(op.getLoc(), range);
+  };
+
+  auto replaceWith = [&](ICmpPredicate predicate, Value lhs,
+                         Value rhs) -> LogicalResult {
+    rewriter.replaceOpWithNewOp<ICmpOp>(op, predicate, lhs, rhs);
+    return success();
+  };
+
+  if (commonPrefixLength == numElements) {
+    // cat(a, b, c) == cat(a, b, c) -> 1
+    switch (op.predicate()) {
+    case ICmpPredicate::ult:
+    case ICmpPredicate::ugt:
+    case ICmpPredicate::ne:
+    case ICmpPredicate::slt:
+    case ICmpPredicate::sgt:
+      return replaceWithConstantI1(0);
+
+    case ICmpPredicate::sle:
+    case ICmpPredicate::sge:
+    case ICmpPredicate::ule:
+    case ICmpPredicate::uge:
+    case ICmpPredicate::eq:
+      return replaceWithConstantI1(1);
+    }
+  } else {
```

Comment:
No need for the "else" after a returning if.  Plz remove it to reduce indentation in the else block

Reply:
Done